source/common/router/rds_impl.cc

```diff
@@ -154,21 +155,24 @@ RouteConfigProviderManagerImpl::routeConfigProviders() {
 };
 
 Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteConfigProvider(
-    const Json::Object& config, Upstream::ClusterManager& cm, Stats::Scope& scope,
+    const envoy::api::v2::filter::Rds& rds, Upstream::ClusterManager& cm, Stats::Scope& scope,
     const std::string& stat_prefix, Init::Manager& init_manager) {
 
-  // RdsRouteConfigProviders are unique based on their <route_config_name>MAP_CONCATENATOR<cluster>.
+  // RdsRouteConfigProviders are unique based on their
+  // <route_config_name>MAP_CONCATENATOR<serialized RDS config>.
+  // TODO(htuch): Full serialization here gives large IDs, could get away with a
+  // strong hash instead.
   const std::string manager_identifier =
-      config.getString("route_config_name") + MAP_CONCATENATOR + config.getString("cluster");
+      rds.route_config_name() + MAP_CONCATENATOR + rds.SerializeAsString();
```

Comment:
In this case, AFAIK unordered_maps don't just hash, they hash and then do a full string compare, so it should be fine. In the API cases, I actually think it's a mistake that we use the hash to decide whether to update or not (though it's nice for performance). Would be good to think about that and see if we want to change that, but that's a separate issue.